### DISCUSSION TITLE
Add temporary effect from outside of effect manager.

### DISCUSSION THREAD

**Discussion by revision29** [PlummersSoftwareLLC/NightDriverStrip#545] [Ideas]
I want to be able to add a temporary effect from outside of effect manager. I propose adding method to the effect manager that effects an effect and then sets that as _tempEffect. Then there is a corresponding method that sets the temp effect to null. Similar to ClearGlobalColor but only touched the temp effect. This would allow temp effects being triggered by an IR remote (such as the DIY1 or FADE3 buttons). I assume it could be helpful for triggering temporary effects from the web interface as well. But, I deal mainly with the IR and don't mess with the web interface as much. I have implemented the following in effectmanager.h in my sandbox:

```
    void SetTemporaryEffect (std::shared_ptr<LEDStripEffect> tempEffect) 
    {
        _tempEffect = tempEffect;
    }

    void ClearTemporaryEffect()
    {
        _tempEffect = nullptr;
    }

```

**Comment by robertlipe** [comment]
Could something like the Memento Design Pattern let you make a stack of
such patterns to push and pop and have checkpoints along the way along with
easy trades without full locking and such?

**Comment by revision29** [comment]
@robertlipe I had to google MDP. I am a simpleton with a severe lack proper code design practices. MDP is a little over my head and I think more complicated than how the effect manager works. I prefer the relatively straightforward functionality.

The reason for this idea is that while I can add an effect to the queue, get it's index number and trigger it, I would much prefer to just set a temporary effect and let it disappear once I am done with it. This can sort of already be gone by applying the global color, but I want to be able to temporarily activate whatever effect I have coded in the IR remote or maybe an effect that I might want to alter push from the web interface.

On the 44 button remote I have, I have 6 DIY buttons. I currently have those mapped to trigger effects 1-6 in the effect manger. But, someone might want those to trigger a special effect outside of the effect queue. Also I am thinking of using FADE3 and FADE7 to implement a palette fill effect that fades through 3 or 7 colors. I need to be able to create and push a temporary effect for that. I need to assume that the effects queue can change at any moment (effects copied, added, or removed) and not hard code references to specific effect indexes.

**Comment by rbergen** [comment]
I think I'd have to see a full use case before I make up my mind about this. 

I'd be particularly interested in how the generic ability to directly set the temporary effect would combine with what you want to use it for - using it to "hard set" specific effects, that may or may not be in the regular effect list. I'd expect the exact effects one wants to trigger to vary from person to person, and I think this would have to be "solved" in a way that is also somewhat generic. 

Then there are things like effect persistence. Currently, the temporary effect is not persisted - it's temporary by nature -, but if we start using it to actually select **the** effect someone wants to look at "forever", that may not be sustainable. But if we start persisting, then what do we do with cases where that doesn't make sense - like with the Mesmerizer splash screen effect?

Again, I think this needs some more "body" before I can say yes or no to it. Maybe you could just implement what you have in mind in a branch in your fork, so we can look at it there?

**Comment by revision29** [comment]
The biggest use case at the moment is if a person actually does want to show the global color RIGHT NOW when they press a color button on their remote. I am glad that we have moved effect manager to its current state of global color, but I know some people might  still want the immediacy of the color showing.

But, overall, I am trying to mimic the abilities of most other generic RGB light strip remote controls, especially the ones with more than 24 buttons. Those usually start showing whatever pre-programmed effect is assigned to a button RIGHT NOW. I am also trying to allow maximum customization of the IR remote experience and bypass the effects queue when desired.

But, there is a greater problem I am trying to avoid. I can accomplish this be simply adding a disabled effect, remembering its index, and then using a button to call that index. But I also know there are other mechanisms for changing the effect manager queue, such as the web interface or receiving API calls. So, I can't explicitly trust that an effect will continue to exist in the queue or stay at its effect index, or that the effect index itself will exist. Telling effect manager to trigger an effect at an index that does not exist causes an uncaught exception and device reboot.

This "solution" is also bad for memory management.

I have been too busy making changes to other areas of the code to make a concrete example, besides the color fill effect. Once the other pull requests get processed, I will start moving my existing projects to my new PersonalProjects branch. There I will have my custom remote control code implemented and that will have examples of what I am trying to accomplish. Those examples can be ported to the stock remote control handling code.

I know this would be waaaaaayyyyyyy down the road, but I also imagine a scenario where the web server might also trigger a temporary effect. Maybe a person previews what an effect might look like before it is added to the queue or before changing an effect in queue. Creativity and real world needs open up the possibilities.

Currently, the only way to set the temporary effect is with the splash screen initializer.

TLDR; I think it is useful to allow a person to set the temporary effect beyond the initializer for the splash effect use case.

**Comment by revision29** [comment]
I will also add this: If a person relies on effects.cpp to intuit what an effect index number is and hard codes that index so they can activate that effect, that assumption might be nullified by a later commit. Example: Dave submitted a PR for changing the order of effects for mesmerizer. If I had set my remote code to activate an effect at index 4, Dave's commit and my subsequent sync with main would mess up which effect is triggered.

**Comment by robertlipe** [comment]
Memento allows stacking backups and restores and implementing from any
point and any limit of depth. The code is pretty close to what
effecrmanager already does, albeit in a strong Json accent. Want to add a
previous as well as a Next? Ok.  Remote or network or timeout or error card
wants to display a new one? Just push/append/emplace it to the stack. It's
not based on effect number.

This is how databases implement rollbacks, editors allow undo, and I on.

I understood that effect numbers were immutable once assigned and couldn't
be changed, even in a code upgrade.

**Comment by rbergen** [comment]
@revision29 The point is that the PR you opened proposes to add functions to a core class in the main tree of project, where the main tree codebase then doesn't use them - that effectively makes the functions dead code on arrival.  
You provide a textual description of a use case that I may or may not understand correctly, but until that's been converted into code itself that's part of the same PR, I can't judge if the overall proposal is consisent and the best approach. That's why I'm holding on what I now consider a partial PR.

Concerning indexes and effect numbers:

- We have an effect number, which is actually an _effect class number_; it was introduced to allow the JSON deserialization logic to figure out which effect class to instantiate for a particular effect JSON blob. A complete deserialized entry in the effect list is thus the effect number (class) plus the effect configuration included in the blob ("constructor parameters").
- The "array index" of an _effect instance_ in the effect list, which will change if an effect instance gets moved within the list.

At some point there was a discussion about adding an effect _instance_ UUID, and I even opened a PR for adding it: #332. I closed it quite a while later because it didn't seem to gather much enthusiasm in the end.

**Comment by revision29** [comment]
I added a use case in the PR, but let me explain my thinking more clearly.

A person might want to have 5 effects that change every 10 minutes or so. Those are added to the effect queue in effects.cpp.

But, a person might want to program their remote to display certain custom effects not in the queue.
A DIY1 button might trigger a fire effect, FADE3 might trigger a palette effect, etc.
The user does not want those in the effect loop, just activate at their command.
So, they modify remote.cpp to map certain buttons to trigger specific temporary effects.

The other way to accomplish this is to add disabled effects to the queue.
They will have to assume in remotecontrol.cpp what the effect index is so they can:

```
effectManager.SetInterval(0);
effectmanager.EnableEffect(index);
effectManager.SetCurrentEfffectIndex(index);
```

The downside to the alternate workaround is that effects are unnecessarily stored in memory and the index number can't be assumed or trusted since index numbers can change dynamically. It is much simpler and less error prone to just define a temporary effect and send it to effect manager since it is transient in nature.

**Reply by rbergen** [reply]
This (i.e., the code you added to the PR) illustrates the conceptual use case, but it's not mergeable into the main tree. The main reasons for that are:

- The selection of the effect in the PR is more or less random, or at least very specific to what you would possibly choose. Another person would almost certainly choose (an)other effect(s). This user-specific flexibility is not taken care of in the PR's current status.
- You're now assigning three functions to two of the remote control buttons, depending on what's the current state: 1) clear global colors if set, 2) clear temporary effect if set, 3) switch to the next or the previous effect. I think that's too much hidden logic behind these buttons to work in practice.

I'm guessing you indeed intended to provide an _example_ of a use case of the kind you have in mind, but until we have a comprehensive "main tree" solution that addresses at least the two issues I just mentioned _and_ relies on the EffectManager functions you propose, there is no real case to add them to the main tree. 

Concerning the "kicking off an effect in response to a color button", that seems a bit contradictory to an earlier conversation we've had in [this comment](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/538#discussioncomment-7651834) in discussion #538. There you made the case to _no longer_ have global colors being set start custom, more or less arbitrary, effects - which is something I agreed with, and still do.

**Reply by revision29** [reply]
> I'm guessing you indeed intended to provide an example of a use case of the kind you have in mind

You asked for an example, and I made one that fits the ethos of the code already it place.

> Concerning the "kicking off an effect in response to a color button"...

The difference is that the user is specifying what effect they want to kick off. Effect manager is not making a choice for them that as coded for a specific use case long ago. I still agree with the route we have taken, but someone might still want to immediately trigger a color (or effect) they specify with a remote button. The code assumes a temporary effect might be set and gives it priority. We have removed the main mechanism for a temp effect to be created. So, this replaces what was lost and creates even more flexibility. This is what I have been trying to achieve from the start: getting EM to stop forcing an effect, and letting me specify what effect or color I want to show RIGHT NOW from the press of a remote.

> You're now assigning three functions to two of the remote control buttons, depending on what's the current state:

True. I am trying to think through how to clear temporary colors and effects with a limited number of buttons. The main branch assumes a 24 button remote. It is very limited, but it still has to be clear out temporary states in addition to do what a user intends. So the button effectively clears temporary states and will then proceed to iterate effects. So it is in concept trying to do two things but requires three actions.

> The selection of the effect in the PR is more or less random, or at least very specific to what you would possibly choose. 

Well, yes, the effect I chose serves as an example of what could be done. Someone has to make the arbitrary choice. As it is someone specified that a user is shown a volume meter. What if I don't have audio enabled? I don't. It will likely crash if I press the button. No matter what, I still have to customize the behavior of that button. ShowVU is a much more limited method in the Effect Manager class that basically establishes a temporary effect (although the effect is stored in a different variable). I want to be able to expand that functionality to any effect, not just the one Dave, or whoever, chose to make an effect manager class for. What I am arguing for is a wider application of a principle already in the code. "Press this button to show this custom effect."

And beyond that, the remote control code out of the box requires heavy customization if a person wants to use a 44 key remote. If I set it to use a 44 key remote, it will not compile. I have been working on a solution for for several months. Hacking and refining as I learn more. A key part of that, no pun intended, will be to trigger a custom effect assigned to a button. The palette effect I chose was chosen to mirror the function of a FADE3 button: fade between three colors. I actually failed because I effectively chose two colors. I got distracted by the sudden jump between color 3 and 1 and just made a wrap. I need to redo it so that it has 3 unique colors with a 4th repeating the first color. Anyway, I want to provide a person with more options with what they can do with the remote control, and not tell them what they should do. Again, the main branch's remote control code is created for a very specific, not end user friendly use case. I want to help improve that. Bottom line, the remote is intended to be customizable and this helps increase its flexibility.

Edit: (to change bad markdown formatting and add silly example below)
Think of this nerdy use case: imagine a Star Trek fan has lights all over their garage. They have a "yellow alert" button and a "red alert" button they want to activate. They spouse walks into the garage. They push the yellow alert button. A temporary color fill effect of yellow is created and displayed RIGHT NOW. She walks away, and the person presses the next button to clear and go back to normal. Or maybe, the person is working on their car and the 10mm socket disappears. That's a red alert. Press the red alert button and all the lights go red. Press the next button when the 10mm is found. In reality, it will always be red because that socket is gone forever. (Context: 10mm is a popular bold head size and the corresponding ratchet socket gets lost a disproportionate number of times). That can't be done right now, nor was it possible to do before. This proposed change allows such silly things to be done.

**Reply by revision29** [reply]
Let me be honest: this seems to be such a simple, useful, and obvious feature to add, I don't understand the pushback. I fully understand not wanting to pollute the codebase with all manner of useless methods. The codebase supports having a tempEffect and gives it drawing priority, but there is no way to set it besides the one use in showing a splashscreen and that is set at initialization. It would be like having the brightness device config option without the ability to set brightness. The variable is there, but no way to meaningfully interact with it. It needs a setter method, otherwise it is useless and should be eliminated from the code. 

The other option is to make tempEffect a publicly accessible variable and let the user interact with it directly. That goes against the design pattern implemented in most of the code and I don't prefer it.

Regardless, this allows a user greater control to do what they want with their remote than whatever arbitrary use cases we have thought of so far. These methods can even be used later to send temp effects from other input sources and no just the remote.

TLDR; a high-priority variable is useless if it can't be set.

**Reply by rbergen** [reply]
**The** objection to adding the accessor functions to the main tree is indeed that there is no use case in the main tree for them. I've also already explained (twice?) that the "personal customization" use case that is now included in #552 doesn't address a number of questions/concerns that a "main codebase" use case has to - and as such doesn't constitute one. 

The fact of the matter is that we already have a **lot** of code to maintain, some of which could do with a clean-up instead of further extensions - and this is something I (and others) have already spent a meaningful amount of time on, as well. In that context I'm just not going to add code to the main tree that is not used in the context of the main tree. Someone involved with this project once said something that relates to all this: in the end, the best code is no code.  
You may think this is overly protective, based too much on principles, or something similar. That's fine, but doesn't change my conclusions or hence decision. 

About the high-priority variable being useless: that's not true as long as there is a main tree use case for _it_. There is and you named it: the splash screen effect. Others may come up later, or the one that's there now may go - in the latter case you can trust that the variable will disappear also.

In the meantime, if you want to support your personal customizations in your fork's branches by adding the functions there that you now propose to add to "main"'s `EffectManager`, you're obviously totally free to do so; I couldn't even stop you if I wanted to. If and when you develop a main tree use case for exposing this variable there, I'm happy to review this again.

**Reply by revision29** [reply]
@rbergen I have pretty much resolved myself to this not making it into the codebase and working to make sure I can patch if a merge eliminates my customization. Making a patch with git is relatively simply. If you don't want to read the wall of text, please skip and read the question at the end.

Essentially what I am asking to think outside of the existing codebase and add a couple of lines of code that can expand the usability of the IR remote. The main branch code assumes a 24 button remote, but there are 44 button remotes with many other buttons that can be assigned to what I am proposing. For example there are sometimes 6 DIY (do it yourself) buttons that would be perfect for this. I can't give an example in code because switching the code to recognize a 44 key remote will cause it to not compile. The code is incomplete. To say it another way, the buttons on a physical remote justify the code addition, not the functionality already present.

I keep mentioning "personal customization" because that is exactly what a person has to do in order to make the IR remote code functional, especially if they use something other than Dave's 24 button remote. Since they already have to customize the remote, adding a really flexible ability of triggering a temporary effect that they specify would allow them to do more than they can now.

The example I provided in the PR takes advantage a button that is only useful for people who have audio enabled. For those who don't have audio enabled, toggling the device setting for showVUMeter does absolutely nothing. Since the button changes what is being shown on the LEDs (if the effect supports it), it is not too far a logical leap to allow a person to instead show an effect they want to show. But, in actuality, the use case I intend is the DIY buttons described above. But, a user can choose to hijack whatever button for this purpose.

Also, here is the biggest main tree justification: we removed code from effect manager that showed a temporary effect. My justification for submitting that PR is that the effect manger should do exactly what the user tells it to do. Setting the global fill color should not trigger a temporary fire effect. We agreed on that. But, the mechanism was there and active for a person to trigger a temporary effect from the remote. That ability is now gone. My draft PR restores and improves on the functionality we removed. With my draft PR, a person can tell effect manager exactly what temporary effect they want to show as they define in the remote control code. And, again, an underlying assumption about the IR code is that a person has to customize it and that customization is encouraged.

The second main tree justification is that the code could support 44 key remotes. Those remotes have do it yourself buttons.  When a person uses such a remote on generic RGB lights, the DIY buttons generally show different effects. That expectation comes with the remote. There is currently no way to replicate that functionality. I can't (currently) help that the main tree remote `handle()` code does not have `else if` statements for the 44 button remote.

The reason I started to contribute to the project in the first place was precisely because I tried to use a 44 key remote and the code failed to build. Then I figured out how to get my 44 key remote working. Then I wanted to take advantage of the extra buttons. I could not figure out how to trigger a temporary effect. I didn't even know how to instantiate a variable of an effect class.  Thought I would come back to it. I got sidetracked by several other issues / fixes. I have finally gotten back to one of the first things I tried to do with the code but could not. Now that I have the knowledge necessary to improve the code by creating additional functionality that adds flexibility. I thought that others could / would benefit from the functionality too. 

I have said and said all of that again, because I don't feel like you understand what I am trying to say. Or, more precisely, I don't feel like I am communicating in a way that leads you to understand me.

ONE LAST QUESTION: would you consider reconsidering if I can provide some use cases of the 44 button remote while also getting the code to not crash when a person chooses the 44 button remote? I won't define all of the buttons, but just enough to provide an example for the user to follow.

**Reply by rbergen** [reply]
What I am apparently not getting across is that I do, and _have to_, consider the codebase as a whole. If code is added within the "overall codebase bubble" that's not used within that same bubble, it's dead code. And I just won't add that, period.

If you feel the PR you submitted earlier removed an effect manager capability that should have stayed, then that should have been considered in the PR in question. I actually expressed hesitation on merging that PR _because_ it limited the impact of global color changes _before_ that was addressed in individual effects, and you successfully argued for pushing forward with it anyway. After you changed my mind on that one I now still think that change constitutes an improvement, so I'm not even considering reverting it.

As I said before, if you or anyone else opens a PR that considers - no, actually, addresses - that:

- People are very likely to have personal preferences about what specific effect gets triggered by which specific remote control button in a setup that directly links buttons to effects, and
- We have configurations controlling WS281x strips, configurations controlling WS281x matrices (potentially of different sizes) and configurations controlling HUB75 matrices (again, possibly of different sizes), all of which use different effect sets and all of which may be IR remote controlled, and
- Effect choice persistence across reboots is now an integrated capability of this project and should be taken care of in some consistent way when people can pick their effect using the remote control, and

...that relies on the functions you were proposing to add to `EffectManager`, I _will_ consider it.

**Reply by revision29** [reply]
This is a really helpful response. The first two bullet points are some of the underlying concepts behind what I am trying accomplish. I have not considered the persistence point since a temporary effect should probably not persist across reboots. But, as I have said before, as it is, the remote code was written for a more specific use case and anyone wanting to use it for their use case has to heavily edit the remotecontrol.cpp file. For me it makes sense for them to also change the code to reflect whatever temporary effect they want to trigger for their use case, using what is there as an example of what they can do. I admit that it is a less than ideal situation.

As I think more about your third bullet point, this is why I want to map an effect to a button. I could add some hacky logic to achieve the same behavior using deactivated effects. But that is unreliable because it requires me to hard code effect index numbers into the IR logic. And I know the index number can be changed due to interactions with effect manager from elsewhere. This would lead to inconsistencies. Mapping a button to a specific temporary effect brings consistency to the experience, regardless of outside changes to effect manager.

One good real world example is for the Christmas tree I made for my front yard. I have several effects in the queue with effect interval at 30000. I might also want to freak the neighbor out and be able to trigger a fire effect that is outside of the effect loop. This can be easily done with deactivated effects in the web interface, but more difficult with the IR remote. So, I might want to assign the DIY button to show a fire effect. Another example: Maybe someone with a matrix has an effect that shows their YT subscriber count with a fancy animation. They might assign that effect to a specific button to show it temporarily. They trigger with a button on their remote, but clear it moments later to return to whatever effect was showing. 

> If you feel the PR you submitted earlier removed an effect manager capability that should have stayed, then that should have been considered in the PR in question.

I don’t think the capability should have stayed in effect manager as it was. But, I find that being able to display a temporary effect of a users own choosing is useful. Showing whatever effect a user maps to a button would be nice. Pressing a color button and being shown a fire effect that was broken was not good. 

> People are very likely to have personal preferences about what specific effect gets triggered by which specific remote control button in a setup that directly links buttons to effects

This is exactly what I am trying to achieve. They would just swap out the effect specified in code with their own. My own remote handling code will have the ability of the user to  associate an effect with a button without having to edit the remotecontrol.cpp file. They just have to edit the button definitions file. But, I’m working within the limitations of the main tree’s IR handling code.

> We have configurations controlling WS281x strips, configurations …

It’s my assumption that all effects are children of the base strip effect class.  Temp effect can be set to whatever effect the user sends, regardless of configuration. My assumption is the ability to temporarily show an effect would be used by anyone on any IR enabled configuration. Am I missing something where this is not the case? 

But, I am going to leave this alone for a while. As you can see, I closed the PR and am going to abandon the idea.  Other things need my attention. Your points above helped me understand more clearly your perspective and triggered some ideas about my own code. So, I’m truly thankful for your input.

**Comment by revision29** [comment]
One last thing for now: after post the walls of text above, my memory kicked back in gear. The 
`if (deviceConfig.ApplyGlobalColors())` logic for IR_BPLUS and IR_BMINUS should actually be removed. I originally had that logic when that one big PR still had effect manager showing the color when applyglobalcolor was called. We needed a way to stop showing it. Now, it does not make sense to clear the global color before going to the next effect. It does make sense to stop the temporary effect and not stop applying the global color. If anything, the power button should clear apply global color on the first press and then on second press act normally. When that happens, it should not clear the global colors, but should only call deviceConfig->ClearApplyGlobalColors(). Just thinking out loud.
